Allow users to specify a set of values that should only be sent to the API

thestrabusiness opened the pull request:
## 🧰 What's being changed?

This commit expands the ability to users to specify a list of keys that should only be sent up to the metrics API. A new filter class was introduced to wrap the `filter_params` and new `allow_only` values and provide the appropriate filter function given the values.

A `HarCollection` wrapper class was also introduced that filters and formats the given hash.

## 🧪 Testing

Using the following curl command and sample app, we confirmed that the application will only send values matching the given list in the headers, query params and bodies. 

`curl 'http://localhost:9292/api/foo?id=1&name=joel' -H "X-Filtered: my custom header" -H "Content-Type: application/json" -X POST -d '{"key": "value", "dont_filter": "not filtered"}`

```ruby
$LOAD_PATH << File.expand_path("../../metrics-sdks/packages/ruby/lib", __FILE__)
require "readme/metrics"

class ApiApp
  def call(env)
    [200, {}, ["Ok"]]
  end
end

map "/api" do
  options = {
    api_key: "YOUR_API_KEY",
    development: false,
    allow_only: ["dont_filter"]
  }
  use Readme::Metrics, options do |env|
    {id: "anthonym", label: "Anthony M.", email: "anthony@example.com"}
  end
  run ApiApp.new
end
```

thestrabusiness commented on packages/ruby/lib/readme/har.rb:
```diff
@@ -67,7 +67,7 @@ def response_body
         begin
           parsed_body = JSON.parse(@response.body.first)
-          return filter_with_params(parsed_body)
+          return HarCollection.new(@filter, parsed_body).to_h.to_json
```
This wouldn't posted to the Metrics API until I called `to_json` on the result of the HarCollection. Should this happen inside the `to_h` method?

JoelQ replied:
No it probably shouldn't convert to json inside `to_h` because I can see wanting to turn that hash into other structures such as URL-encoded string or XML

JoelQ replied:
We disovered that we we do indeed want to double-encode JSON

JoelQ commented on packages/ruby/lib/readme/filter.rb:
```diff
@@ -0,0 +1,43 @@
+    end
+
+    def filter(hash)
+      hash.select { |key, _value| !@filter_values.include?(key) }
```
Thoughts on using `reject` here instead of negating the conditional?

JoelQ commented on packages/ruby/lib/readme/har.rb:
```diff
@@ -67,7 +67,7 @@ def response_body
         begin
           parsed_body = JSON.parse(@response.body.first)
-          return filter_with_params(parsed_body)
+          return HarCollection.new(@filter, parsed_body).to_h.to_json
```
Do we need an explicit `return` here?

thestrabusiness replied:
We do need this to return early out of the `response_body` method. Otherwise this will always return `@response.body.each.reduce(:+)` on line 77.

JoelQ replied:
Right, my suggestion was to convert:

```ruby
if condition
  return value
end

default
````

into the more explicit conditional and avoiding the inline return:

```ruby
if conditional
  value
else
  default
end
```

thestrabusiness replied:
Ah, got it. Will do 👍

JoelQ commented on packages/ruby/lib/readme/har_request.rb:
```diff
@@ -1,18 +1,18 @@
 module Readme
   class HarRequest
-    def initialize(request, filter_params = [])
+    def initialize(request, filter = Filter::None.new)
```
Is the default necessary here? I think we always return `Filter::None.new` when calculating a filter in the first place?

JoelQ commented on packages/ruby/lib/readme/har_request.rb:
```diff
@@ -26,26 +26,9 @@ def as_json
     rescue
       @request.body
     end
-
-    def to_hash_array(hash)
-      hash.map { |name, value| {name: name, value: value} }
-    end
-
-    def filter_hash_with_params(hash)
-      hash.select { |key, _value| !@filter_params.include?(key) }
-    end
-
-    def filter_hash_array_with_params(array)
-      array.select { |pair| !@filter_params.include? pair[:name] }
-    end
-
-    def filtered_hash_array(hash)
-      as_array = to_hash_array(hash)
-      filter_hash_array_with_params(as_array)
-    end
```
Love to see all this red!

JoelQ commented on packages/ruby/lib/readme/metrics.rb:
```diff
@@ -1,5 +1,7 @@
 require "readme/metrics/version"
 require "readme/har"
+require "readme/filter"
+require "readme/har_collection"
```
I don't think we need the har collection in this file?

JoelQ commented on packages/ruby/spec/readme/har_collection_spec.rb:
```diff
@@ -0,0 +1,77 @@
+  describe "#to_h" do
+    context "with FilterParams filter" do
+      it "returns the given hash without the given filter values" do
+        filter_values = ["filter"]
+        filter = Filter::FilterParams.new(filter_values)
```
Thoughts on combining these two lines? At some point too many variables have the name "filter" in them and it's hard to keep them straight 😅 

```ruby
filter = Filter::FilterParams.new(["filter"])
```

JoelQ commented on packages/ruby/spec/readme/har_collection_spec.rb:
```diff
@@ -0,0 +1,77 @@
+      it "returns the given hash without the given filter values" do
+        filter_values = ["filter"]
+        filter = Filter::FilterParams.new(filter_values)
+        hash = {"filter" => "filtered", "dont_filter" => "not filtered"}
```
Thoughts on giving these names like "keep" and "reject"? At some point too many variables have the name "filter" in them and it's hard to keep them straight 😅

thestrabusiness replied:
Heh can do.

JoelQ commented on packages/ruby/spec/readme/har_request_spec.rb:
```diff
@@ -66,10 +66,10 @@
       json = request.as_json
 
-      request_body = json.dig(:postData, :text)
+      request_body = JSON.parse(json.dig(:postData, :text))
```
Hmmm is this asserting that the body is doubly encoded?

thestrabusiness replied:
I'm not sure about this. When I introduced the `HarCollection` object, I was unable to send data to the Metrics API with our sample application. Adding the `to_json` call on the request body fix that issue, and so I updated this test accordingly. It seems like introducing the `HarCollection` changed something, but its not clear what.

JoelQ replied:
I ran into this exact same issue on my branch. Even the schema was complaining. Not sure how it was not failing spec before 🤷

JoelQ commented on packages/ruby/spec/readme/har_spec.rb:
```diff
@@ -104,13 +105,13 @@
 
       response = json.dig("log", "entries", 0, "response")
-      response_body = response["content"]["text"]
+      response_body = JSON.parse(response["content"]["text"])
```
Again, I think this is saying that the body is doubly encoded, which we don't want?

JoelQ commented on packages/ruby/lib/readme/filter.rb:
```diff
@@ -0,0 +1,43 @@
+class Filter
+  def self.for(filter_params = nil, allow_only = nil)
```
Thoughts on making these keyword args? Something like:

```ruby
def self.for(reject: nil, keep_only: nil)
```

JoelQ commented on packages/ruby/lib/readme/filter.rb:
```diff
@@ -0,0 +1,43 @@
+class Filter
+  def self.for(filter_params = nil, allow_only = nil)
+    if !filter_params.nil? && !allow_only.nil?
+      raise "Can only supply either filter_params or allow_only"
+    end
+
+    unless filter_params.nil?
+      return FilterParams.new(filter_params)
+    end
+
+    unless allow_only.nil?
+      return AllowOnly.new(allow_only)
+    end
+
+    None.new
```
Thoughts on making a single 4-way conditional here instead of relying on early returns?

JoelQ commented:
The word "filter" is getting a bit overloaded in here. Thoughts on avoiding it when talking about sensitive values being rejected? Particularly since "filter" tends to be ambiguous in whether it's selecting or rejecting items.

JoelQ commented:
I love the strategy approach to the filter objects. Thoughts on unit testing the various objects in `filter.rb` in their own spec, separately from the har collection? That way we don't need to test that the har collection works with all the different filtering permutations, just that given a filter it applies it and returns the proper hash or array. You could even do something with a double like:

```ruby
describe "#to_hash" do
  it "returns a hash with filtering applied" do
    data = { keep: "keep", reject: "reject" }
    filter = double(filter: {keep: "keep"})
    collection = HarCollection.new(filter, data)

    expect(collection.to_h).to eq { keep: "keep" }
  end
end
```

thestrabusiness commented:
> I love the strategy approach to the filter objects. Thoughts on unit testing the various objects in `filter.rb` in their own spec, separately from the har collection? That way we don't need to test that the har collection works with all the different filtering permutations, just that given a filter it applies it and returns the proper hash or array. You could even do something with a double like:
> 
> ```ruby
> describe "#to_hash" do
>   it "returns a hash with filtering applied" do
>     data = { keep: "keep", reject: "reject" }
>     filter = double(filter: {keep: "keep"})
>     collection = HarCollection.new(filter, data)
> 
>     expect(collection.to_h).to eq { keep: "keep" }
>   end
> end
> ```

Good call. I addressed your comments and split out the tests as you described.

JoelQ commented on packages/ruby/spec/readme/filter_spec.rb:
```diff
@@ -0,0 +1,59 @@
+  describe ".for" do
+    it "returns RejectParams when only reject argument is given" do
+      result = Filter.for(reject: ["reject"])
+      expect(result.class).to eq Filter::RejectParams
+    end
+
+    it "returns AllowOnly when only allow_only argument is given" do
+      result = Filter.for(allow_only: ["keep"])
+      expect(result.class).to eq Filter::AllowOnly
+    end
+
+    it "returns None when neither arugment is given" do
+      result = Filter.for
+      expect(result.class).to eq Filter::None
+    end
```
A slightly RSpec way to write these expectations would be:

```ruby
expect(result).to be_an_instance_of Filter::None
```

thestrabusiness replied:
Nice, I'll update this
